### PR TITLE
Caret location improvements 

### LINKED ISSUE #92080: The cursor height becomes shorter when I delete all characters from TextField
like the following picture:

<details>
<summary>flutter doctor -v</summary>

```log
[√] Flutter (Channel stable, 2.5.3, on Microsoft Windows [Version 10.0.17763.2237], locale    en-US)
    • Flutter version 2.5.3 at D:\SDK\Flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • FLUTTER_GIT_URL = https://gitee.com/mirrors/Flutter.git
    • Framework revision 18116933e7 (4 days ago), 2021-10-15 10:46:35 -0700
    • Engine revision d3ea636dc5
    • Dart version 2.14.4
    • Pub download mirror https://pub.flutter-io.cn
    • Flutter download mirror https://storage.flutter-io.cn

[√] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at D:\SDK\Android
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = D:\SDK\Android
    • ANDROID_SDK_ROOT = D:\SDK\Android
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • CHROME_EXECUTABLE = C:\Program Files (x86)\Microsoft\Edge\Application\msedge.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.5)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.11.31729.503
    • Windows 10 SDK version 10.0.19041.0

[√] Android Studio (version 4.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
       https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
       https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[√] IntelliJ IDEA Community Edition (version 2021.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2021.2.3
    • Flutter plugin version 61.2.4
    • Dart plugin version 212.5486

[√] VS Code (version 1.55.2)
    • VS Code at C:\Users\Zhanghuifeng\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.15.0

[√] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.17763.2237]
    • Chrome (web)      • chrome  • web-javascript • unknown
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 94.0.992.50

• No issues found!
```
</details>

<details>
<summary>simple code</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'TextField Demo',
      theme: ThemeData(primarySwatch: Colors.teal),
      home: Scaffold(
        body: Center(
          child: SizedBox(
            width: 200,
            height: 80,
            child: TextField(
              controller: TextEditingController(text: "Hello"),
              minLines: 1,
              textAlignVertical: TextAlignVertical.center,
              cursorHeight: 14,
              decoration: InputDecoration(
                border: const OutlineInputBorder(),
                hintText: "HintText",
                labelText: "LabelText",
                hintStyle: const TextStyle(fontSize: 14),
                labelStyle: const TextStyle(fontSize: 15),
                counterText:  "",
              ),
            ),
          ),
        ),
      ),
    );
  }
}
```
</details>

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
- Fixes https://github.com/flutter/flutter/issues/92080
- Removes some usages of `toPlainText`, as it could be expensive.
- `getLocalRectForCaret` now reports the real rect that will be painted.
- move caret x-coordinate clamping to RenderEditable since TextPainter doesn't know about clipping.

## Pre-launch Checklist
 
- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by LongCatIsLooong** [comment]
4.1K scuba changes but nothing stood out.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 1]
```diff
@@ -895,188 +895,253 @@ class TextPainter {
...
-      return Rect.fromLTRB(clampDouble(dx, 0, _paragraph!.width), box.top,
-          clampDouble(dx, 0, _paragraph!.width), box.bottom);
+  // Computes the _CaretMetrics for the text within the half-open interval
+  // [start, end).
+  //
+  // This call can be expensive.
+  //
+  // Returns null when the range [start, end) does not contain a full grapheme
+  // in the paragrah.
+  static _CaretMetrics? _caretMetricsFromTextBox(int start, int end, TextAffinity affinity, ui.Paragraph paragraph) {
```
[the remark this answers is not in the document]
Comment (justinmc) [reply packages/flutter/lib/src/painting/text_painter.dart:919]: I see it's used to indicate direction elsewhere in this PR.  Maybe it's a standard thing or maybe it requires a little bit more description in these comments.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 2]
```diff
@@ -895,188 +895,253 @@ class TextPainter {
...
+    // line break, force the search direction to be downstream. This is because
+    // newline characters are reported as zero-width TextBoxes at the **end** of
+    // its line. The behavior of different getBoxesForRange implementations
+    // is not consistent when the input range is [length, length + 1) and the
+    // last character is a line breaker.
+    //
+    // TODO(LongCatIsLooong): introduce ui.Paragraph-level api for this. The
+    // exponential search using getBoxesForRange isn't necessary and does not
+    // work well for certian edge cases (e.g., when the position is inside a
+    // glyph, and the previous character is a newline character).
```
Comment (justinmc) [inline packages/flutter/lib/src/painting/text_painter.dart:1068]: Nit: Consider linking/creating a GitHub issue here.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 3]
```diff
@@ -765,17 +767,16 @@ class TextPainter {
...
           newWidth = _applyFloatingPointHack(_paragraph!.longestLine);
           break;
         case TextWidthBasis.parent:
           newWidth = maxIntrinsicWidth;
           break;
       }
+      // TODO(LongCatIsLooong): ui.Paragraph.width always reports the input
+      // width. The double layout keeps the two width in sync so TextPainter
+      // does not have to work with a non-zero paint offset. SkParagraph
+      // invalidates the layout entirely in this case so this is expensive.
```
Comment (justinmc) [inline packages/flutter/lib/src/painting/text_painter.dart:779]: Should we create an issue for this?  Or maybe it's not possible or easy to reproduce a userland bug caused by this.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 4]
```diff
@@ -895,188 +895,253 @@ class TextPainter {
...
     }
-    _caretMetrics = _CaretMetrics(
-      offset: rect != null ? Offset(rect.left, rect.top) : _emptyOffset,
-      fullHeight: rect != null ? rect.bottom - rect.top : null,
-    );
-
-    // Cache the input parameters to prevent repeat work later.
-    _previousCaretPosition = position;
-    _previousCaretPrototype = caretPrototype;
+    return _cachedCaretMetrics = caretMetrics;
```
Comment (justinmc) [inline packages/flutter/lib/src/painting/text_painter.dart:1144]: Why did you change this to both return and assign?

**Review by justinmc — COMMENTED** [review]
Do you know what's going on with the scuba changes?  I see that you changed a few pixel values in the tests.  Should the caret height be the only thing that changed visually?

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 2 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:1068]: This will be taken care of in the new paragraph api. But I don't think there's a github issue.

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:779]: This is one important use case that has to be covered in the new paragraph API. I'll describe that in the design doc instead.

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 4 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:1144]: caretPrototype doesn't affect the anchoring of the caret and it's not expensive at all to compute, there's no need to invalidate the cache over that.

**Comment by guidezpl** [comment]
I think this also fixes https://github.com/flutter/flutter/issues/64072

**Comment by justinmc** [comment]
@LongCatIsLooong Are you still planning to merge this?  Sorry I never gave a final review.  Looks like it needs to be merged with master now.

**Comment by LongCatIsLooong** [comment]
I'll get back to this once needed lower-level APIs are exposed in ui.Paragraph.
